hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -721,24 +722,23 @@ def __attrs_post_init__(self):
 class ConjectureData(object):
     @classmethod
     def for_buffer(self, buffer, observer=None):
-        buffer = hbytes(buffer)
         return ConjectureData(
-            max_length=len(buffer),
-            draw_bytes=lambda data, n: hbytes(buffer[data.index : data.index + n]),
-            observer=observer,
+            prefix=buffer, max_length=len(buffer), parameter=None, observer=observer,
         )
 
-    def __init__(self, max_length, draw_bytes, observer=None):
+    def __init__(self, max_length, prefix, parameter, observer=None):
         if observer is None:
             observer = DataObserver()
         assert isinstance(observer, DataObserver)
+        self.__bytes_drawn = 0
         self.observer = observer
         self.max_length = max_length
         self.is_find = False
-        self._draw_bytes = draw_bytes
         self.overdraw = 0
         self.__block_starts = defaultdict(list)
         self.__block_starts_calculated_to = 0
+        self.__prefix = prefix
+        self.__parameter = parameter
```

Comment:
Since `parameter` can be `None` for buffer-only inputs, this assertion will probably come in handy at some point:

```python
assert parameter is not None or max_length <= len(prefix)
```